File: src/main/kotlin/com/s4ng/jobbylanguage/service/WebCrawlerService.java
package com.s4ng.jobbylanguage.service;

import com.gargoylesoftware.htmlunit.WebClient;
import com.gargoylesoftware.htmlunit.html.HtmlButton;
import com.gargoylesoftware.htmlunit.html.HtmlPage;
import com.s4ng.jobbylanguage.model.dto.JobOpeningDto;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class WebCrawlerService {

	public List< JobOpeningDto > getDataFromWeb() throws IOException {

		WebClient webClient = new WebClient();
		webClient.getOptions().setThrowExceptionOnScriptError( false );
		webClient.getOptions().setJavaScriptEnabled( true );
		webClient.getOptions().setRedirectEnabled( true );
		java.util.logging.Logger.getLogger( "com.gargoylesoftware" ).setLevel( Level.OFF );

		HtmlPage page = webClient.getPage( "https://www.sara" + "min.co.kr/zf_user/jobs/list/job-category" );
		webClient.waitForBackgroundJavaScript( 50000 );

		HtmlButton itButton = page.getFirstByXPath( "//*[@id=\"sp_main_wrapper\"]/div[2]/div/div[1]/div[2]/div[1]/button[6]" );
		itButton.click();
		webClient.waitForBackgroundJavaScript( 50000 );

		List< HtmlButton > buttonList = new ArrayList<>();

		int count = 1;
		HtmlButton jobNameButton;
		do {
			jobNameButton = page.getFirstByXPath( "//*[@id=\"sp_job_category_subDepth_2\"]/div[2]/div/div[2]/div/dl[3]/dd/button[" + count + "]" );
			if ( jobNameButton != null ) {
				buttonList.add( jobNameButton );
			}
			count++;
		} while ( jobNameButton != null );

		return buttonList.stream().map( HtmlButton::asNormalizedText )
				.map( this::toDto )
				.collect( Collectors.toList() );
	}

	private JobOpeningDto toDto( String text ) {

		String name;
		String count;

		String[] tokens = text.split( "\\(" );

		name = tokens[ 0 ];
		count = tokens[ 1 ].substring( 0, tokens[ 1 ].length() - 1 ).replaceAll( ",", "" );

		return new JobOpeningDto( name, Integer.parseInt( count ) );
	}
}

File: src/main/kotlin/com/s4ng/jobbylanguage/service/WebCrawlerService.kt
package com.s4ng.jobbylanguage.service

import com.gargoylesoftware.htmlunit.Page
import com.gargoylesoftware.htmlunit.WebClient
import com.gargoylesoftware.htmlunit.html.HtmlButton
import com.gargoylesoftware.htmlunit.html.HtmlPage
import com.s4ng.jobbylanguage.model.dto.JobOpeningDto
import org.springframework.stereotype.Service
import java.io.IOException
import java.util.Objects
import java.util.logging.Level
import java.util.logging.Logger
import java.util.stream.Collectors

@Service
class WebCrawlerService {

    fun getDataFromWeb(): MutableList<JobOpeningDto> {

        val webClient = WebClient()
        webClient.options.isThrowExceptionOnScriptError = false
        webClient.options.isJavaScriptEnabled = true
        webClient.options.isRedirectEnabled = true

        Logger.getLogger("com.gargoylesoftware").level = Level.OFF
        val page = webClient.getPage<HtmlPage>("https://www.sara" + "min.co.kr/zf_user/jobs/list/job-category")
        webClient.waitForBackgroundJavaScript(50000)

        val itButton = page.getFirstByXPath<HtmlButton>("//*[@id=\"sp_main_wrapper\"]/div[2]/div/div[1]/div[2]/div[1]/button[6]")
        itButton.click<Page>()
        webClient.waitForBackgroundJavaScript(50000)

        val buttonList: MutableList<HtmlButton> = ArrayList()
        var count = 1
        var jobNameButton: HtmlButton?
        do {
            jobNameButton = page.getFirstByXPath("//*[@id=\"sp_job_category_subDepth_2\"]/div[2]/div/div[2]/div/dl[3]/dd/button[$count]")
            if (jobNameButton != null) {
                buttonList.add(jobNameButton)
            }
            count++
        } while (jobNameButton != null)

        return buttonList.stream().map { obj: HtmlButton -> obj.asNormalizedText() }
                .map { text: String -> toDto(text) }
                .collect(Collectors.toList())
    }

    private fun toDto(text: String): JobOpeningDto? {

        val name: String
        val count: String
        val tokens = text.split("\\(".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        name = tokens[0]
        count = tokens[1].substring(0, tokens[1].length - 1).replace(",".toRegex(), "")
        return JobOpeningDto(name, count.toInt())
    }
}